src/tags/tests/test_models.py

```diff
@@ -0,0 +1,276 @@
+        self.assertEqual(total_xp_by_tags(self.user, unpacked_tag_list), sum(xp_list))
+
+    def test_one_tag_multiple_badges_only(self):
+        """ 
+            See if correct xp is displayed using only one tag + badges with multiple badgeassertions
+        """
+        xp_list = self.xp_list[:len(self.set_list)]  # 10
+
+        for i in range(len(xp_list)):
+            badge, ba = self.create_badge(xp_list[i], self.set_list[i])
+            badge.tags.add("TAG")
+
+        self.assertEqual(total_xp_by_tags(self.user, ["TAG"]), sum(xp_list[i] * self.set_list[i] for i in range(len(self.set_list))))
+
+    def test_one_tag_badge_only__different_semesters(self):
+        """ 
+            See if correct xp is displayed using only one tag + badges with multiple badgeassertions in different semesters
+        """ 
+        semester_set = [SiteConfig().get().active_semester] + baker.make("courses.semester", _quantity=2)
+        for semester in semester_set:
+            badge, ba = self.create_badge(100)
+            badge.tags.add("TAG")
+
+            ba[0].semester = semester
+            ba[0].save()
+
+        self.assertEqual(total_xp_by_tags(self.user, ["TAG"]), 100)
+
+    # QUEST + BADGE TESTS
+
+    def test_one_tag_quests_badges(self):
+        """ 
+            See if correct xp is displayed using only 1 tag added to Quest and Badge objects
+        """ 
+        xp_list = self.xp_list[:20]  # 20
+        
+        for i in range(len(xp_list) // 2):
+            quest, _ = self.create_quest(next(self.xp_iter))
+            quest.tags.add("TAG")
+
+            badge, _ = self.create_badge(next(self.xp_iter))
+            badge.tags.add("TAG")
+
+        self.assertEqual(total_xp_by_tags(self.user, ["TAG"]), sum(xp_list))
+
+    def test_multiple_separate_tags_quests_badges(self):
+        """ 
+            See if correct xp is displayed using only multiple tags added to Quest and Badge objects
+            only 1 unique tag per quest and badge
+        """ 
+        tag_list = [f"tag-{i}" for i in range(5 * 2)]  # 10
+        xp_list = self.xp_list[:5 * len(tag_list)]  # 50
+
+        for tag_el in tag_list:
+            for i in range(len(xp_list) // len(tag_list)):
+
+                if i % 2 == 0:
+                    quest, _ = self.create_quest(next(self.xp_iter))
+                    quest.tags.add(tag_el)
+                else:
+                    badge, _ = self.create_badge(next(self.xp_iter))
+                    badge.tags.add(tag_el)
+
+        self.assertEqual(total_xp_by_tags(self.user, tag_list), sum(xp_list))
+
+    def test_multiple_crossing_tags_quest_badges(self):
+        """ 
+            See if correct xp is displayed using only multiple tags added to Quest and Badge objects
+            each Quest and Badge object should have multiple tags
+        """ 
+        tag_list = [(f"tag0-{i}", f"tag1-{i}") for i in range(5 * 2)]
```

Comment:
Do this in a python prompt, or print the output, then paste the literal values here explicitly.

Also, maybe call it `tag_tuples` which is more accurate (plural variable names usually indicate lists, so you want the variable to tell you what it's a list of)